asserts/extkeypairmgr.go

```diff
@@ -0,0 +1,260 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+
+	"golang.org/x/crypto/openpgp/packet"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+// ExternalKeypairManager is key pair manager implemented via an external program interface.
+// TODO: points to interface docs
+type ExternalKeypairManager struct {
+	keyMgrPath string
+	nameToID   map[string]string
+	cache      map[string]*cachedExtKey
+}
+
+// NewExternalKeypairManager creates a new ExternalKeypairManager using the program at keyMgrPath.
+func NewExternalKeypairManager(keyMgrPath string) (*ExternalKeypairManager, error) {
+	em := &ExternalKeypairManager{
+		keyMgrPath: keyMgrPath,
+		nameToID:   make(map[string]string),
+		cache:      make(map[string]*cachedExtKey),
+	}
+	if err := em.checkFeatures(); err != nil {
+		return nil, err
+	}
+	return em, nil
+}
+
+func (em *ExternalKeypairManager) keyMgr(op string, args []string, in []byte, out interface{}) error {
+	args = append([]string{op}, args...)
+	cmd := exec.Command(em.keyMgrPath, args...)
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	if len(in) != 0 {
+		cmd.Stdin = bytes.NewBuffer(in)
+	}
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("external keypair manager %q %v failed: %v (%q)", em.keyMgrPath, args, err, errBuf.Bytes())
+
+	}
+	switch o := out.(type) {
+	case *[]byte:
+		*o = outBuf.Bytes()
+	default:
+		err := json.Unmarshal(outBuf.Bytes(), out)
+		if err != nil {
+			return fmt.Errorf("cannot decode external keypair manager %q %v output: %v", em.keyMgrPath, args, err)
+		}
+	}
+	return nil
+}
+
+func (em *ExternalKeypairManager) checkFeatures() error {
+	var feats struct {
+		Signing    []string `json:"signing"`
+		PublicKeys []string `json:"public-keys"`
+	}
+	if err := em.keyMgr("features", nil, nil, &feats); err != nil {
+		return err
+	}
+	if !strutil.ListContains(feats.Signing, "RSA-PKCS") {
+		return fmt.Errorf("external keypair manager %q unexpectedly does not support RSA-PKCS", em.keyMgrPath)
+	}
+	if !strutil.ListContains(feats.PublicKeys, "DER") {
+		return fmt.Errorf("external keypair manager %q unexpectedly does not support public key DER output format", em.keyMgrPath)
```

Comment:
I'd drop "unexpectedly", these errors look a bit weird; maybe @degville has some input?

Reply:
I agree that "unexpectedly" does look odd in this context, although I can see why it's there. It is clearer dropping it, or perhaps use _missing_? eg. `external keypair manager %q missing support for the public key DER output format`

Reply:
these are the basic features without which the external keypair manager is useless, but I agree "unexpectedly" sounds odd, I was probably too concentrated on my mental model of this when I wrote the errors. "missing support" sounds good